core/src/main/scala/kafka/server/DelayedOperation.scala

```
@@ -370,6 +383,20 @@ final class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: Stri
 
       purged
     }
+
+    def forceComplete(): Int = {
```

The current behavior in DelayedProduce is that if forceComplete() is called, all pending partitions will return a  REQUEST_TIMED_OUT error. However, in the case when the coordinator has changed, we should return a NOT_COORDINATOR error instead.

---

Any clue as to how i would check this in the `DelayedProduce.onComplete` ?

---

One way to do this is to first mark the coordinator as not available, then simply call Purgatory.checkAndComplete(). Those pending produce requests are guaranteed to complete since those pending partitions will error out after checking the coordinator state in tryComplete(). Then, we don't need this forceComplete() call. We probably need to call Purgatory.checkAndComplete() while holding the coordinator lock so that we can guarantee that the coordinator state doesn't change when the purgatory does the check.

---

Actually, we can't hold the leader lock while checking the purgatory. Another way to do that is to remember the expected leader epoch in the delayed produce request. After the leader changes, the leader epoch will be advanced. If the delayed produce request errors out on non-matching leader epoch, then a tryComplete() call on the purgatory is guaranteed to complete all outstanding requests.

---

Actually looking into `Partition.checkEnoughReplicasReachOffset` if we `tryComplete` they  should error out with `Errors.NOT_LEADER_FOR_PARTITION`. This will be returned as `Errors.NOT_COORDINATOR` from the TC. Perhaps that will do?

---

Looking further into it - i think this already works without having to do anything. Correct me if i'm wrong, but when a leadership change happens `ReplicaManager.makeFollowers` will get called. Which does:
```
partitionsToMakeFollower.foreach { partition =>
        val topicPartitionOperationKey = new TopicPartitionOperationKey(partition.topicPartition)
        tryCompleteDelayedProduce(topicPartitionOperationKey)
        tryCompleteDelayedFetch(topicPartitionOperationKey)
      }
```
At which point the `DelayedProduce` will error out with `Errors.NOT_LEADER_FOR_PARTITION` which will be translated to `Errors.NOT_COORDINATOR` in `TransactionStateManger.appendTransactionToLog`. So I think there is perhaps nothing to do

---

Good point. Then, I agree this is already covered.